python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -1188,8 +1202,20 @@ def _iterative_groupby_apply(
         elif isinstance(chunk_results[0], cudf.Series) and isinstance(
             self.obj, cudf.DataFrame
         ):
-            result = cudf.concat(chunk_results, axis=1).T
-            result.index.names = self.grouping.names
+            # When the UDF is like df.sum(), the result for each
+            # group is a row-like "Series" where the index labels
+            # are the same as the original calling DataFrame
+            if _is_row_of(chunk_results[0], self.obj):
+                result = cudf.concat(chunk_results, axis=1).T
+                result.index = group_names
+                result.index.names = self.grouping.names
+            # When the UDF is like df.x + df.y, the result for each
+            # group is the same length as the original group
+            elif len(self.obj) == sum(len(chk) for chk in chunk_results):
```

Comment:
There's no `else` block here. Can we write this in a way that prevents `result` from being an unbound variable? Even if the `else` block is raising a descriptive error, it's better than letting `result` be an unbound variable that is needed later in the function.